plugins/linux-capture/xcomposite-helpers.cpp

```diff
@@ -0,0 +1,568 @@
+#include <glad/glad.h>
+#include <glad/glad_glx.h>
+#include <X11/Xlib.h>
+#include <xcb/xcb.h>
+#include <xcb/composite.h>
+#include <pthread.h>
+
+#include <list>
+#include <map>
+#include <unordered_set>
+#include <cstdint>
+
+#include <util/platform.h>
+#include "xcomposite-input.hpp"
+
+// Atoms used throughout our plugin
+xcb_atom_t ATOM_UTF8_STRING;
+xcb_atom_t ATOM_STRING;
+xcb_atom_t ATOM_TEXT;
+xcb_atom_t ATOM_COMPOUND_TEXT;
+xcb_atom_t ATOM_WM_NAME;
+xcb_atom_t ATOM_WM_CLASS;
+xcb_atom_t ATOM__NET_WM_NAME;
+xcb_atom_t ATOM__NET_SUPPORTING_WM_CHECK;
+xcb_atom_t ATOM__NET_CLIENT_LIST;
+
+xcb_atom_t get_atom(xcb_connection_t *conn, const char *name)
+{
+	auto atom_c = xcb_intern_atom(conn, 1, strlen(name), name);
+	auto atom_r = xcb_intern_atom_reply(conn, atom_c, nullptr);
+	xcb_atom_t a = atom_r->atom;
+	free(atom_r);
+	return a;
+}
+
+void xcomp_gather_atoms(xcb_connection_t *conn)
+{
+	ATOM_UTF8_STRING = get_atom(conn, "UTF8_STRING");
+	ATOM_STRING = get_atom(conn, "STRING");
+	ATOM_TEXT = get_atom(conn, "TEXT");
+	ATOM_COMPOUND_TEXT = get_atom(conn, "COMPOUND_TEXT");
+	ATOM_WM_NAME = get_atom(conn, "WM_NAME");
+	ATOM_WM_CLASS = get_atom(conn, "WM_CLASS");
+	ATOM__NET_WM_NAME = get_atom(conn, "_NET_WM_NAME");
+	ATOM__NET_SUPPORTING_WM_CHECK =
+		get_atom(conn, "_NET_SUPPORTING_WM_CHECK");
+	ATOM__NET_CLIENT_LIST = get_atom(conn, "_NET_CLIENT_LIST");
+}
+
+// XErrorHandler for glXCreatePixmap calls.
+static bool pixmap_err = false;
+static char pixmap_err_text[200];
+static int catch_pixmap_errors(Display *display, XErrorEvent *err)
+{
+	pixmap_err = true;
+	memset(pixmap_err_text, 0, 200);
+	XGetErrorText(display, err->error_code, pixmap_err_text, 200);
+	return 0;
+}
+
+int xcb_get_screen_for_root(xcb_connection_t *conn, xcb_window_t root)
+{
+	xcb_screen_iterator_t iter =
+		xcb_setup_roots_iterator(xcb_get_setup(conn));
+	for (int i = 0; iter.rem > 0; xcb_screen_next(&iter)) {
+		if (iter.data->root == root)
+			return i;
+		i++;
+	}
+	return 0;
```

Comment:
It is written to never fail, so it will always return a valid screen. For valid input that will be the screen associated with the root id, for invalid input we just give you screen 0 which always exists. Avoiding callers having to error check and modeling the Xlib implementation of this function that its standing in for.

Reply:
Fair enough.